File: 백준/Gold/2169. 로봇 조종하기/로봇 조종하기.py
import sys

read = sys.stdin.readline

N, M = map(int, read().split())

board = []
for _ in range(N):
    tmp = list(map(int, read().split()))
    board.append(tmp)

# 첫 행은 왼쪽에서 오른쪽으로 가는 방법밖에 없음
for i in range(1):
    for j in range(1, M):
        board[i][j] += board[i][j-1]

for i in range(1, N):
    # 행별로 위쪽에서 내려오는 경우
    up = [0 for _ in range(M)]
    for j in range(M):
        up[j] = board[i-1][j]

    # 왼 -> 오 경우들의 가치
    left = [0 for _ in range(M)]
    left[0] = up[0] + board[i][0]
    for j in range(1, M):
        # 왼쪽에서 오는 경우, 위에서 내려오는 경우 중 최대값을 찾기
        left[j] = max(left[j - 1], up[j]) + board[i][j]

    # 오 -> 왼 경우들의 가치
    right = [0 for _ in range(M)]
    right[M-1] = up[M-1] + board[i][M-1]
    for j in range(M - 2, -1, -1):
        # 오른쪽에서 오는 경우, 위에서 내려오는 경우 중 최대값을 찾기
        right[j] = max(right[j + 1], up[j]) + board[i][j]

    # 왼 -> 오, 오 -> 왼 중 최대 가치로 현재 위치를 갱신
    for j in range(M):
        board[i][j] = max(left[j], right[j])

print(board[N-1][M-1])

File: 백준/Gold/2169. 로봇 조종하기/로봇 조종하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(bf.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[][] board = new int[N][M];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(bf.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int[][] dp = new int[N][M];

        // 첫 행은 왼쪽에서만 올 수 있음
        dp[0][0] = board[0][0];
        for (int i = 1; i < M; i++) {
            dp[0][i] = dp[0][i - 1] + board[0][i];
        }

        // 이후 각 행의 첫행 or 끝행은 위에서 내려오는 값
        for (int i = 1; i < N; i++) {
            int[] left = new int[M];
            int[] right = new int[M];

            // 좌 -> 우
            left[0] = dp[i-1][0] + board[i][0];
            for (int j = 1; j < M; j++) {
                // 위쪽 -> 아래 or 왼쪽 -> 오른쪽으로 오는 것 중 젤 큰 값
                left[j] = Math.max(left[j-1], dp[i-1][j]) + board[i][j];
            }

            // 우 -> 좌
            right[M-1] = dp[i-1][M-1] + board[i][M-1];
            for (int j = M-2; j >= 0; j--) {
                // 위쪽 -> 아래 or 오른쪽 -> 왼쪽으로 오는 것 중 젤 큰 값
                right[j] = Math.max(right[j+1], dp[i-1][j]) + board[i][j];
            }

            // 둘 중 최댓 값 선택
            for(int j=0; j<M; j++){
                dp[i][j] = Math.max(left[j], right[j]);
            }
        }
        System.out.println(dp[N-1][M-1]);
    }
}
